File: 2019_KAKAO_BLIND_RECRUITMENT/Open_Chatroom/Main.java
import java.util.HashMap;
import java.util.ArrayList;

public class Main {
	public static void main(String [] args) {
		Solution s = new Solution();
		String [] record = {"Enter uid1234 Muzi", "Enter uid4567 Prodo","Leave uid1234","Enter uid1234 Prodo","Change uid4567 Ryan"};
		String [] n = s.solution(record);
	}
}

class Solution {
	HashMap<String, String> map;
	ArrayList<Pair> arr;

	public String[] solution(String[] record) {
        String[] answer;
		
		map = new HashMap<>();
		arr = new ArrayList<>();

		for (int i = 0 ; i < record.length ; i++) {
			String line = record[i];
			String token[] = line.split(" ");
			String op, userId, nickname = "null";
				
			op = token[0];
			userId = token[1];

			if (token.length == 3) nickname = token[2];

			if (op.equals("Enter")) {
				enter_op(userId, nickname);
			} else if (op.equals("Leave")) {
				arr.add(new Pair(nickname, "Leave", userId));
			} else if (op.equals("Change")) {
				map.put(userId, nickname);
			}
		}	
	
		answer = buildAns();	
        return answer;
    }

	public String [] buildAns() {
		String [] ans = new String[arr.size()];
		for (int i = 0 ; i < arr.size() ; i++) {
			Pair p = arr.get(i);
			String line = "";
			String userId = p.id;	
			String nickname = map.get(userId);	
			line += nickname;
			if (p.op.equals("Enter")) {
				line += "님이 들어왔습니다.";
			} else if (p.op.equals("Leave")) {
				line += "님이 나갔습니다.";
			} 
			ans[i] = line;
			//System.out.println(line);
		}
		return ans;
	}

	public void enter_op(String userId, String nickname) { 
		arr.add(new Pair(nickname, "Enter", userId));
		map.put(userId,nickname);
	}

	class Pair {
		String nickname, op, id;
		public Pair(String nick, String op, String id) {
			this.nickname = nick;
			this.op = op;
			this.id = id;
		}
	}
}

File: 2019_KAKAO_BLIND_RECRUITMENT/Open_Chatroom/Main.js
class Pair {
    constructor (userId, op) {
        this._userId = userId;
        this._op = op;
    }
    get getUserId() { return this._userId; }
    get getOp() { return this._op; }
}

function solution(record) {
    var answer = [];

    let map = new Map();
    let array = new Array();

    for (let i = 0 ; i < record.length ; i++) {
        let line = record[i];
        let tokens = line.split(' ');
        let op , userId, nickname;
        op = tokens[0];
        userId = tokens[1];
        if (tokens.length == 3) { nickname = tokens[2]; }
        
        if ( op == "Enter") {
            array.push(new Pair(userId, op));
            map.set(userId, nickname);
        } else if ( op == "Leave") {
            array.push(new Pair(userId, op));
        } else if (op == "Change") {
            map.set(userId, nickname);
        }
    }

    for (let i = 0 ; i < array.length ; i++) {
        let line = "";
        let pair = array[i];
        let userId = pair.getUserId;
        let operation = pair.getOp;
        let nickname = map.get(userId);

        line += ( nickname + "님이 ");
        if (operation == "Enter") {
            line += "들어왔습니다.";
        } else if (operation == "Leave") {
            line += "나갔습니다.";
        }
        answer.push(line);
    }
    
    return answer;
}



